File: arrays/2_pilas.js
// #region Pila_JS
class Pila {
    #items = [];

    push(item) {
        this.#items.push(item);
    }

    pop() {
        if (this.isEmpty()) return "Underflow";
        return this.#items.pop();
    }

    peek() {
        return this.#items[this.#items.length - 1];
    }

    isEmpty() {
        return this.#items.length === 0;
    }
}

// Aplicación: Validar si una cadena de texto está balanceada
function verificarParentesis(cadena) {
    const pila = new Pila();
    for (let char of cadena) {
        if (char === '(') {
            pila.push(char);
        } else if (char === ')') {
            if (pila.isEmpty()) return false;
            pila.pop();
        }
    }
    return pila.isEmpty();
}

console.log(verificarParentesis("(())")); // true
console.log(verificarParentesis("(()"));  // false
// #endregion

File: avance/arrays/2_pilas.py
# #region Pila_Pro
class Pila:
    """Implementación de una Pila usando el principio LIFO."""
    def __init__(self):
        self.items = []

    def esta_vacia(self):
        return not self.items

    def push(self, item):
        # O(1) Amortizado
        self.items.append(item)

    def pop(self):
        # O(1)
        if self.esta_vacia():
            raise IndexError("La pila está vacía")
        return self.items.pop()

    def peek(self):
        """Ver el elemento superior sin sacarlo."""
        return self.items[-1]

# Uso del algoritmo: Invertir una palabra
def invertir_texto(texto):
    pila = Pila()
    for char in texto:
        pila.push(char)
    
    resultado = ""
    while not pila.esta_vacia():
        resultado += pila.pop()
    return resultado

print(invertir_texto("Estructura")) # arutcurtsE
# #endregion